### PR TITLE
Implement whois

### PR DESCRIPTION
**Pull request by DavidSpenler**
### Pull Request Checklist

* [x] I have added any new tests that need to pass to `sytest-whitelist` as specified in [docs/sytest.md](https://github.com/matrix-org/dendrite/blob/master/docs/sytest.md)
* [x] Pull request includes a [sign off](https://github.com/matrix-org/dendrite/blob/master/docs/CONTRIBUTING.md#sign-off)

Signed-off-by: `David Spenler <dsspenle@uwaterloo.ca>`

### CODE REVIEW

**Comment by tulir** [comment]
What's the point of this without admin accounts? (I don't know why the endpoint is in the spec at all)

**Comment by DavidSpenler** [comment]
This PR addresses #1306. It gets us closer to spec compliance regardless of how useful it is at present.

File: clientapi/routing/admin_whois.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+			"": deviceInfo{
+				Sessions: []sessionInfo{
+					sessionInfo{
+						Connections: connections,
+					},
+				},
+			},
+		},
+	}
+
```
Comment (S7evinK) [inline clientapi/routing/admin_whois.go:88]: As per the spec, "Each key is an identifier for one of the user's devices."
Synapse just uses an empty string, not sure if we should do the same. (Well, if the `device.ID` is empty, we'll get the same result)

As i had started on this endpoint, when i extended the table with `LastSeenIP` etc.:
```suggestion
	res := make(map[string]deviceInfo)
	for _, device := range devicesResp.Devices {
		connInfo := connectionInfo{
			IP:        device.LastSeenIP,
			LastSeen:  device.LastSeenTS,
			UserAgent: device.UserAgent,
		}
		dev, ok := res[device.ID]
		if !ok {
			dev.Sessions = []sessionInfo{{}}
		}
		dev.Sessions[0].Connections = append(dev.Sessions[0].Connections, connInfo)
		res[device.ID] = dev
	}
```

**Comment by DavidSpenler** [comment]
We've made the changes you suggested.

**Review by neilalexander — COMMENTED** [review]
Do you have a plan for protecting this endpoint? We don't currently have a concept of admin accounts and this looks like it could expose user IP addresses at current.

**Comment by DavidSpenler** [comment]
The endpoint in its current state does not allow the user to query anyone other than themselves. We assumed that the additional permission of admins to query anyone would have to wait until the admin functionality itself is implemented.

**Comment by DavidSpenler** [comment]
Can I get an update on this?

**Review by neilalexander — APPROVED** [review]
We had a bit of discussion about this since we don't have any concept of admin accounts and/or plans for other unspecced admin endpoints, but since this one is in the spec and seems to do what the spec says, we will merge it.

Other unspecced admin endpoints are going to be excluded via matrix-org/sytest#981 so they won't affect our test compliance.

**Pull request merged**